src/rust/engine/fs/src/snapshot.rs

```diff
@@ -130,53 +130,81 @@ impl Snapshot {
+      })
+      .to_boxed()
   }
 
-  fn contents_for_directory_sync(
+  // Assumes that all fingerprints it encounters are valid.
+  // Returns an unsorted Vec.
+  fn contents_for_directory_helper(
     fingerprint: Fingerprint,
     store: Arc<Store>,
     path_so_far: PathBuf,
-  ) -> Result<Vec<FileContent>, String> {
-    let directory = store
+  ) -> BoxFuture<Vec<FileContent>, String> {
+    store
       .load_directory_proto(fingerprint)
-      .wait()
-      .unwrap()
-      .ok_or_else(|| format!("Could not find snapshot {}", fingerprint))?;
-
-    let mut contents = Vec::new();
-
-    for file_node in directory.get_files() {
-      contents.push(FileContent {
-        path: path_so_far.join(file_node.get_name()),
-        content: store
-          .load_file_bytes(
-            Fingerprint::from_hex_string(file_node.get_digest().get_hash()).unwrap(),
-          )
-          .wait()
-          .unwrap()
-          .unwrap(),
-      });
-    }
-
-    for dir_node in directory.get_directories() {
-      contents.extend(
-        Snapshot::contents_for_directory_sync(
-          Fingerprint::from_hex_string(dir_node.get_digest().get_hash()).unwrap(),
-          store.clone(),
-          path_so_far.join(dir_node.get_name()),
-        ).unwrap(),
-      );
-    }
-
-    contents.sort_by(|a, b| a.path.cmp(&b.path));
-    Ok(contents)
+      .and_then(move |maybe_dir| {
+        maybe_dir.ok_or_else(|| {
+          format!("Could not find directory with fingerprint {}", fingerprint)
+        })
+      })
+      .and_then(move |dir| {
+        let file_futures = join_all(
+          dir
+            .get_files()
+            .iter()
+            .map(|file_node| {
+              let path = path_so_far.join(file_node.get_name());
+              let maybe_bytes =
+                store.load_file_bytes(
+                  Fingerprint::from_hex_string(file_node.get_digest().get_hash()).unwrap(),
+                );
+              futures::future::ok(path).join(maybe_bytes)
+            })
+            .collect::<Vec<_>>(),
+        );
+        let dir_futures = join_all(
+          dir
+            .get_directories()
+            .iter()
+            .map(|dir_node| {
+              Snapshot::contents_for_directory_helper(
+                Fingerprint::from_hex_string(dir_node.get_digest().get_hash()).unwrap(),
+                store.clone(),
+                path_so_far.join(dir_node.get_name()),
+              )
+            })
+            .collect::<Vec<_>>(),
+        );
+        file_futures.join(dir_futures)
+      })
+      .and_then(
+        move |(paths_and_maybe_byteses, dirs): (Vec<(PathBuf, Option<Vec<u8>>)>,
+                                                Vec<Vec<FileContent>>)| {
+          join_all(
+            paths_and_maybe_byteses
+              .into_iter()
+              .map(|(path, maybe_bytes)| {
+                maybe_bytes
+                  .ok_or_else(|| format!("Couldn't find file contents for {:?}", path))
```

Comment:
Seems like handling this error above would simplify the type of `paths_and_maybe_byteses` to just `Vec<FileContent>`... any reason to defer it to this separate `join_all` call?